src/dkg/round1.rs

```diff
@@ -0,0 +1,167 @@
+use crate::frost::keys::VerifiableSecretSharingCommitment;
+use crate::frost::Field;
+use crate::frost::Identifier;
+use crate::frost::JubjubScalarField;
+use crate::multienc;
+use crate::multienc::MultiRecipientBlob;
+use crate::participant;
+use crate::serde::read_u16;
+use crate::serde::read_variable_length;
+use crate::serde::write_u16;
+use crate::serde::write_variable_length;
+use rand_core::CryptoRng;
+use rand_core::RngCore;
+use std::io;
+use std::mem;
+
+type Scalar = <JubjubScalarField as Field>::Scalar;
+
+/// Copy of the [`frost_core::dkg::round1::SecretPackage`] struct. Necessary to implement
+/// serialization for this struct. This must be kept in sync with the upstream version.
+struct SerializableSecretPackage {
+    identifier: Identifier,
+    coefficients: Vec<Scalar>,
+    commitment: VerifiableSecretSharingCommitment,
+    min_signers: u16,
+    max_signers: u16,
+}
+
+impl SerializableSecretPackage {
+    fn serialize_into<W: io::Write>(&self, mut writer: W) -> io::Result<()> {
+        writer.write_all(&self.identifier.serialize())?;
+        write_variable_length(&mut writer, &self.coefficients, |writer, scalar| {
+            writer.write_all(&scalar.to_bytes())
+        })?;
+        write_variable_length(&mut writer, self.commitment.serialize(), |writer, array| {
+            writer.write_all(&array)
+        })?;
+        write_u16(&mut writer, self.min_signers)?;
+        write_u16(&mut writer, self.max_signers)?;
+        Ok(())
+    }
+
+    fn deserialize_from<R: io::Read>(mut reader: R) -> io::Result<Self> {
+        let mut identifier = [0u8; 32];
+        reader.read_exact(&mut identifier)?;
+        let identifier = Identifier::deserialize(&identifier).map_err(io::Error::other)?;
+
+        let coefficients = read_variable_length(&mut reader, |reader| {
+            let mut scalar = [0u8; 32];
+            reader.read_exact(&mut scalar)?;
+            let scalar: Option<Scalar> = Scalar::from_bytes(&scalar).into();
+            scalar.ok_or_else(|| io::Error::other("coefficients deserialization failed"))
+        })?;
+
+        let commitment = VerifiableSecretSharingCommitment::deserialize(read_variable_length(
+            &mut reader,
+            |reader| {
+                let mut array = [0u8; 32];
+                reader.read_exact(&mut array)?;
+                Ok(array)
+            },
+        )?)
+        .map_err(io::Error::other)?;
+
+        let min_signers = read_u16(&mut reader)?;
+        let max_signers = read_u16(&mut reader)?;
+
+        Ok(Self {
+            identifier,
+            coefficients,
+            commitment,
+            min_signers,
+            max_signers,
+        })
+    }
+}
+
+impl From<SecretPackage> for SerializableSecretPackage {
+    #[inline]
+    fn from(pkg: SecretPackage) -> Self {
+        // SAFETY: The fields of `SecretPackage` and `SerializableSecretPackage` have the same
+        // size, alignment, and semantics
+        unsafe { mem::transmute(pkg) }
```

Comment:
wondering if it makes sense to define these immediately after the struct definition? since their safety depends on `SerializableSecretPackage` having the same structure as `SecretPackage`